src/connections/mongodb.ts

```diff
@@ -0,0 +1,301 @@
+        }
+
+        await this.client.connect()
+        this.db = this.client.db(this._.dbName)
+    }
+
+    /**
+     * Performs a disconnect action on the current Connection object.
+     *
+     * @returns Promise<any>
+     */
+    async disconnect(): Promise<any> {
+        return this.client?.close()
+    }
+
+    /**
+     * Triggers a query action on the current Connection object.
+     *
+     * The parameters object is sent along with the query to be used in the
+     * query. By default if the query has parameters the SQL statement will
+     * produce a string with `?::[DataType]` values that the parameters object
+     * keys should map to, and will be replaced by.
+     *
+     * @param query - The query to be executed.
+     * @returns Promise<{ data: any, error: Error | null, query: string }>
+     */
+    async query(
+        query: Query
+    ): Promise<{ data: any; error: Error | null; query: string }> {
+        const db = this.db
+        if (!db) throw new Error('No MongoDB connection was found.')
+
+        let result = null
+        let error = null
+        let rawSQL = null
+
+        try {
+            const { res } = await this.runQuery(query.query)
+            rawSQL = constructRawQuery(query)
+            result = res
+        } catch (e) {
+            error = e instanceof Error ? e : new Error(String(e))
+            rawSQL = constructRawQuery(query)
+        }
+
+        return {
+            data: result,
+            error: error,
+            query: rawSQL,
+        }
+    }
+
+    public async fetchDatabaseSchema(): Promise<Database> {
+        if (!this.db) throw new Error('No MongoDB connection was found.')
+
+        const collections = await this.db.listCollections().toArray()
+
+        const schemaMap: Record<string, Record<string, Table>> = {}
+
+        for (const collectionInfo of collections) {
+            const collectionName = collectionInfo.name
+            const collection = this.db.collection(collectionName)
+            const indexes = await collection.indexes()
+
+            // Since MongoDB is schemaless, we sample a document to infer the schema
+            const sampleDocument = await collection.findOne()
+            const columns: TableColumn[] = []
+
+            if (sampleDocument) {
+                let position = 0
+                for (const [fieldName, value] of Object.entries(
+                    sampleDocument
+                )) {
+                    const column: TableColumn = {
+                        name: fieldName,
+                        type: typeof value,
+                        position: position++,
+                        nullable: true,
+                        default: null,
+                        primary: fieldName === '_id',
+                        unique: fieldName === '_id',
+                        references: [],
+                    }
+                    columns.push(column)
+                }
+            }
```

Comment:
Not knowing much about Mongo or the tools it provides, no. I can do some digging. The lowest hanging fruit would be to sample multiple documents. Unless there's a way to find all columns in a collection that's easy, you're going to need to check every document within a collection.

Reply:
@Brayden 
To make sure we have less reads and not require a sample size which isn't a guarantee. We can refetch the schema on every pagination call if we wanted to. Just change the frontend to make its own schema based on the data that's returned. 
```
            {
                "_id": "5db1c37e4a68c31f10cf0a9f",
                "name": "foo",
                "email": "foobaz@bar.com",
                "password": "foobar",
                "preferences": {}
            },
            {
                "_id": "66fafaebb1dd96af381fbf5d",
                "banana": "cheese"
            }
        ],
        ```
        The API will give you all of the column names available like this on the UI but this will cause problems with the structure tab, which is probably something else that needs to be talked about for NoSQL options